src/compiler/checker.ts

```diff
@@ -13592,12 +13592,13 @@ namespace ts {
             return result;
         }
 
-        function createIndexedAccessType(objectType: Type, indexType: Type, aliasSymbol: Symbol | undefined, aliasTypeArguments: readonly Type[] | undefined) {
+        function createIndexedAccessType(objectType: Type, indexType: Type, aliasSymbol: Symbol | undefined, aliasTypeArguments: readonly Type[] | undefined, shouldIncludeUndefined: boolean) {
             const type = <IndexedAccessType>createType(TypeFlags.IndexedAccess);
             type.objectType = objectType;
             type.indexType = indexType;
             type.aliasSymbol = aliasSymbol;
             type.aliasTypeArguments = aliasTypeArguments;
+            type.noUncheckedIndexedAccessCandidate = shouldIncludeUndefined;

```

Comment:
Maybe this should just be named `shouldIncludeUndefined`?

Reply:
I did that at first, but I felt like it was confusing because that’s only true if the type resolves to an index signature. Indexed access types often resolve to an actual property, where they should _not_ include undefined. Happy to hear pithier suggestions that don’t imply the unconditional presence of `undefined` though.